File: src/main/kotlin/com/draw/drawmusic/history/TrackHistory.java
package com.draw.drawmusic.history;

import com.draw.drawmusic.track.Track;
import com.draw.drawmusic.track.TrackBar;

import java.util.function.Consumer;

public class TrackHistory extends Behaviors{
    Track object;

    private static final Consumer<Track> add = TrackBar::addElement;
    private static final Consumer<Track> delete = TrackBar::deleteElement;
    private static final Consumer<Track> moveUp = TrackBar::moveUp;
    private static final Consumer<Track> moveDown = TrackBar::moveDown;

    private final Consumer<Track> undoMethod;
    private final Consumer<Track> redoMethod;

    protected TrackHistory(Track object, Consumer<Track> undoMethod, Consumer<Track> redoMethod) {
        super();
        this.object = object;
        this.undoMethod = undoMethod;
        this.redoMethod = redoMethod;
    }

    @Override
    protected void undoMethod() {
        undoMethod.accept(object);
    }

    @Override
    protected void redoMethod() {
        redoMethod.accept(object);
    }

    public static void saveAddHistory(Track object) {
        new TrackHistory(object, delete, add);
    }

    public static void saveDeleteHistory(Track object) {
        new TrackHistory(object, add, delete);
    }

    public static void saveMoveUpHistory(Track object) {
        new TrackHistory(object, moveDown, moveUp);
    }

    public static void saveMoveDownHistory(Track object) {
        new TrackHistory(object, moveUp, moveDown);
    }
}

File: src/main/kotlin/com/draw/drawmusic/history/TrackHistory.kt
package com.draw.drawmusic.history

import com.draw.drawmusic.track.Track
import com.draw.drawmusic.track.TrackBar
import java.util.function.Consumer

class TrackHistory private constructor(
    private val track: Track,
    private val undoMethod: Consumer<Track>,
    private val redoMethod: Consumer<Track>
) :
    Behaviors() {
    override fun undoMethod() {
        undoMethod.accept(track)
    }

    override fun redoMethod() {
        redoMethod.accept(track)
    }

    companion object {
        private val add = TrackBar::addElement
        private val delete = TrackBar::deleteElement
        private val moveUp = TrackBar::moveUp
        private val moveDown = TrackBar::moveDown

        fun saveAddHistory(`object`: Track) {
            TrackHistory(`object`, delete, add)
        }

        fun saveDeleteHistory(`object`: Track) {
            TrackHistory(`object`, add, delete)
        }

        fun saveMoveUpHistory(`object`: Track) {
            TrackHistory(`object`, moveDown, moveUp)
        }

        fun saveMoveDownHistory(`object`: Track) {
            TrackHistory(`object`, moveUp, moveDown)
        }
    }
}
